### DISCUSSION TITLE
analyze vs run & decode [FEATURE REQUEST] ?

### DISCUSSION THREAD

**Discussion by sbilello** [github/codeql#5333] [Q&A] [answered]
When I run the analyze command I get an error message INVALID_RESULT_PATTERNS while if I breakdown the execution and the result decoding it works correctly. Why?
```
❯ codeql database analyze \
  --rerun  \
  --format=csv \
  --output=output.csv \
  propdb \
  ./properties.ql
Running queries.
Compiling query plan for /Users/sergiobilello/Documents/repositories/asset-service/properties.ql.
[1/1] Found in cache: /Users/sergiobilello/Documents/repositories/asset-service/properties.ql.
Starting evaluation of asset-service/properties.ql.
[1/1 eval 217ms] Evaluation done; writing results to asset-service/properties.bqrs.
Shutting down query evaluator.
Interpreting results.
A fatal error occurred: Could not process query metadata.
Error was: Expected result pattern(s) are not present for query kind "problem": Expected alternating entity/string and string columns. [INVALID_RESULT_PATTERNS]
```

VS 
I created a script `prop.sh` to automate the steps for running and results decoding

❯ ./prop.sh properties.ql
+ codeql query run --database propdb properties.ql --output=constants.bqrs
Compiling query plan for /Users/sergiobilello/Documents/repositories/asset-service/properties.ql.
[1/1] Found in cache: /Users/sergiobilello/Documents/repositories/asset-service/properties.ql.
Starting evaluation of asset-service/properties.ql.
[1/1 eval 221ms] Evaluation done; writing results to /Users/sergiobilello/Documents/repositories/asset-service/constants.bqrs.
Shutting down query evaluator.
+ codeql bqrs decode --output=decoded.csv --format=csv constants.bqrs
+ cat decoded.csv
"col0","col1","col2"
"/Users/sergiobilello/Documents/repositories/asset-service/.mvn/wrapper/maven-wrapper.properties:1[1-115]","distributionUrl","https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip"
"/Users/sergiobilello/Documents/repositories/asset-service/.mvn/wrapper/maven-wrapper.properties:2[1-101]","wrapperUrl","https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/0.5.6/maven-wrapper-0.5.6.jar"
```

**Comment by sbilello** [comment, marked as answer]
I understand that analyze has strict metadata restrictions accordingly to the documentation but it can not be added a flag to avoid that you need to write your own script to get the results in the format that the developer requests? 
I also found this problem https://github.com/github/codeql/issues/5319

**Reply by hvitved** [reply]
The meta data on the query (in this case `@kind problem`) requires that the output has a certain structure, so either you have to change the output structure to be valid, or use `codeql query run` as you have done above.

**Reply by aeisenberg** [reply]
Marking this as answered, please let us know if there is anything left that is unclear.
